#84 Nuxt tutorial

evelinabe opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to appwrite here: https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md

Happy contributing!

-->

## What does this PR do?

This PR adds a tutorial to the website for an idea tracker with Nuxt and Appwrite

## Test Plan


## Related PRs and Issues

Related PRs and Issues
https://github.com/appwrite/website/issues/84

Repo with the complete code for the tutorial: https://github.com/evelinabe/nuxt-ideas-tracker-tutorial
### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes

gewenyu99 commented:
@evelinabe is this ready for review?

evelinabe commented:
@gewenyu99 yes

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-2/+page.markdoc:
```diff
@@ -0,0 +1,107 @@
+As an additional step, by adding [Appwrite's Pink Design system]('https://pink.appwrite.io/') we will also have a global layout system to help with the layout.
+
+
+### 1. Add a page
```
Please don't use numbers in headings.

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-2/+page.markdoc:
```diff
@@ -0,0 +1,107 @@
+---
+layout: tutorial
+title: Build an ideas tracker with Nuxt
```
Why is this title the same as the first page?

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-2/+page.markdoc:
````diff
@@ -0,0 +1,107 @@
+npm run dev
+```
+
+# Pages and layouts for routing {% #pages-and-layout-for-routing %}
````
Can we make this sub heading short and concise? This is a little long

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-2/+page.markdoc:
```diff
@@ -0,0 +1,107 @@
+
+### 2. Add a default layout
+
+ a file `src/layouts/default.vue` and insert the following code:
```
Please make sure you're writing complete sentences. Use a period instead of `:`

evelinabe replied:
Removed

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-2/+page.markdoc:
````diff
@@ -0,0 +1,107 @@
+```
+
+
+### 4. Import layout
````
These sub headings should just be part of the sentences. I don't really understand their value :)

evelinabe replied:
Removed

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-4/+page.markdoc:
```diff
@@ -0,0 +1,173 @@
+
+# Nuxt composables {% #nuxt-composables %}
+
+Composables is a pattern in Nuxt to manage logic related to data fetching and global state management.
```
We need just a little more context. I feel like this thought is incomplete, when do we use composables? Why are we creating one here?

gewenyu99 replied:
How is this related to authentication? Make sure to explain all of this so it makes sense to readers.

evelinabe replied:
Added some more context throughout the tutorial.

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-4/+page.markdoc:
````diff
@@ -0,0 +1,173 @@
+
+```
+
+# User section on home page {% #user-section-on-home-page %}
````
Can this be more concise? This heading is massive

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-4/+page.markdoc:
```diff
@@ -0,0 +1,173 @@
+step: 4
+---
+
+# Nuxt composables {% #nuxt-composables %}
```
Just Composables.

It's a nuxt tutorial, no need to repeat the domain.

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-5/+page.markdoc:
```diff
@@ -0,0 +1,51 @@
+---
+
+In our app we want to show a navigation bar. We will add it to the `app.vue` component and use the `useUserSession` composable to show either:
+- a logout button if the user is logged in
```
No need for a list here. Just write it as a sentence.

evelinabe replied:
Paragraph rewritten.

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-5/+page.markdoc:
```diff
@@ -0,0 +1,51 @@
+- a logout button if the user is logged in
+- or a login button if the user is not logged in
+
+Update the `app.vue` file:
```
With what? What changed, why are we making this change? Can we maybe also add comments on the parts that changed in the code snippet?

evelinabe replied:
Comments added in code throughout the tutorial, explanations added.

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-6/+page.markdoc:
```diff
@@ -0,0 +1,72 @@
+| title       | String | Yes      |
+| description | String | No       |
+
+Change the collection's permissions settings to give any role access to it.
```
Can we use a more specific role and add instructions?

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-7/+page.markdoc:
```diff
@@ -0,0 +1,129 @@
+---
+layout: tutorial
+title: Show ideas on home page
```
Can this be a more concise title that's more consistent with other tutorials?

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-7/+page.markdoc:
```diff
@@ -0,0 +1,129 @@
+step: 7
+---
+
+# Show and add ideas {% #show-and-add-ideas %}
```
Don't repeat the title in a heading.

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-7/+page.markdoc:
```diff
@@ -0,0 +1,129 @@
+              <h5 class="heading-level-6">{{ idea.title }}</h5>
+              <p class="body-text-2">{{ idea.description }}</p>
+              <div
+                class="u-position-absolute u-inset-inline-end-8 u-inset-block-start-8"
```
This code block is realllllyyy massive. Can it be more concise? Can we break it up?

evelinabe replied:
I broke it up in smaller pieces. Started with the form, moved on to the list and ended with the small styling tag.

gewenyu99 reviewed (changes requested):
Hey, great start! Thank you so much for helping us improve Appwrite's docs.

I've left some comments on changes, many of the comments apply to **the entire tutorial** so be sure to carefully apply them across the tutorial, so we can speed up the review times.

Thanks!

gewenyu99 commented:
@evelinabe remember to re-request review when you've finished ur changes!

evelinabe commented:
@gewenyu99 Hi! Requesting re-review. Changed the tone of the texts, please let me know what you think and if you want some editorial changes to the texts. I'm not native in English, so all tips are welcome.

gewenyu99 commented:
@evelinabe Hi can you fix your builds?

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-2/+page.markdoc:
````diff
@@ -0,0 +1,98 @@
+Install the JavaScript Appwrite SDK.
+
+```sh
+npm install appwrite
````
missing `cd` command.

evelinabe replied:
Added

evelinabe commented:
> @evelinabe Hi can you fix your builds?

I hope I fixed it, removed a file with changes. 🤞

gewenyu99 commented:
> > @evelinabe Hi can you fix your builds?
> 
> Strange, thought I had removed the file causing the errors from the PR, but still failure 🧐 @gewenyu99

Please run pnpm run build.

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-4/+page.markdoc:
```diff
@@ -0,0 +1,196 @@
+    login,
+    logout,
+    register,
+  }
```
Missing a line here

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-4/+page.markdoc:
```diff
@@ -0,0 +1,196 @@
+
+# Home page {% #home-page %}
+
+Now that users can register and login to the app, let's modify the home page in `src/pages/index.vue` to give the logged in users a secret message.
```
Didn't we just change this in the last step??

Are the files you refer to correct?

gewenyu99 commented:
I tried following this tutorial but it was really confusing. I couldn't tell which file we were in, what we were implementing, and when you used `...` to show skipped code, it was very confusing to see what was skipped and where the code should go.

Please take a break, and take a careful look through your tutorial from scratch. Follow your own tutorial, it'll help you iron out the kinks :D

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-6/+page.markdoc:
```diff
@@ -0,0 +1,79 @@
+import { database } from "~/appwrite";
+import { ref } from "vue";
+
+const IDEAS_DATABASE_ID = "YOUR_DATABASE_ID";
```
This should be lifted to a config file.

evelinabe replied:
Normally I'd have tucked this away in the`.env`-file, however that's not how you've constructed the other tutorials. 

This file is the equivalent of the `store` in SvelteKit
or `context`in React and follows their pattern.

gewenyu99 replied:
I would still use a `.env` file.

gewenyu99 replied:
I think the other tutorials should follow this pattern too :) 

Not conflicting. We're still improving existing tutorials, too :D

evelinabe replied:
`.env` file added 👯

gewenyu99 commented:
I would take a look at existing tutorials.

Always explain what's going to happen in a step first.

Clearly indicate new files and changes.

If possible, split the files and avoid showing a file in multiple steps, it is that much more difficult to follow.

If you must, use empty placeholder tags and functions, and clearly indicate them as we'll replace later.

Label your code snippets with comments at the top to show which file we're in.

evelinabe commented:
> I would take a look at existing tutorials.
> 
> Always explain what's going to happen in a step first.
> 
> Clearly indicate new files and changes.
> 
> If possible, split the files and avoid showing a file in multiple steps, it is that much more difficult to follow.
> 
> If you must, use empty placeholder tags and functions, and clearly indicate them as we'll replace later.
> 
> Label your code snippets with comments at the top to show which file we're in.

Just want to point out that when I followed the other tutorials you thought that that major code block on step 7 was to big so this was my attempt to divide it in smaller bites. 

A difference is of course that the styling takes up a lot of space. It's always confusing to me with tutorials that first shows an image on what will be built but then ends up totally unstyled. It is a really disappointing thing as a beginner to get a totally different result than what I'm expecting from the first image, so I wanted to keep the styling.

I also added more explicit functions in the script sections than in the other tutorials to make it clearer what functionality is actually going on. But it takes more space.
 
In a real world project I'd of course have a component for the ideaform and another one for the list. But that would mean moving even further from the existing tutorials.

I'm getting the feeling that you on the one hand want me to follow the other tutorials, but if I do that you want me to to impove where they fall short. And this is an Hacktober issue after all.

I will go through it once more. I'm inclined to restore step-7 to that first major block of code to make it follow the other tutorials, though.

evelinabe commented:
@gewenyu99 I have marked all the files and restructured the code a bit. Decided to add UI components to break the code up more 'naturally'. So there are more files, but I hope it's easier to follow.

gewenyu99 commented:
> > I would take a look at existing tutorials.
> > Always explain what's going to happen in a step first.
> > Clearly indicate new files and changes.
> > If possible, split the files and avoid showing a file in multiple steps, it is that much more difficult to follow.
> > If you must, use empty placeholder tags and functions, and clearly indicate them as we'll replace later.
> > Label your code snippets with comments at the top to show which file we're in.
> 
> Just want to point out that when I followed the other tutorials you thought that that major code block on step 7 was to big so this was my attempt to divide it in smaller bites.
> 
> A difference is of course that the styling takes up a lot of space. It's always confusing to me with tutorials that first shows an image on what will be built but then ends up totally unstyled. It is a really disappointing thing as a beginner to get a totally different result than what I'm expecting from the first image, so I wanted to keep the styling.
> 
> I also added more explicit functions in the script sections than in the other tutorials to make it clearer what functionality is actually going on. But it takes more space.
> 
> In a real world project I'd of course have a component for the ideaform and another one for the list. But that would mean moving even further from the existing tutorials.
> 
> I'm getting the feeling that you on the one hand want me to follow the other tutorials, but if I do that you want me to to impove where they fall short. And this is an Hacktober issue after all.
> 
> I will go through it once more. I'm inclined to restore step-7 to that first major block of code to make it follow the other tutorials, though.

This tutorial is inherently more complex, so we will try to make improvements where applicable :)

Better is the only way forward ;)

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-6/+page.markdoc:
```diff
@@ -0,0 +1,88 @@
+| description | String | No       |
+
+Change the collection's permissions in the settings to give access.
+First, add the role "Any" and check the `READ` box.
```
Applies to all, in general in docs, we don't use quotes to specify things we refer to in UI.

For anything you're quoting from UI, use **bold**.

For anything you quote from a file or is copy-able, use a `code-tags`

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-7/+page.markdoc:
````diff
@@ -0,0 +1,222 @@
+```
+
+Next, add the component to the page `pages/index.vue` by auto-importing it in the `<template>`tag.
+In doing that, we need to take a moment to think of how we want to display the form to our.
````
I think there's some words missing in this sentence.

gewenyu99 commented on src/routes/docs/tutorials/nuxt/step-7/+page.markdoc:
```diff
@@ -0,0 +1,222 @@
+
+Next, add the component to the page `pages/index.vue` by auto-importing it in the `<template>`tag.
+In doing that, we need to take a moment to think of how we want to display the form to our.
+Since it should only be shown to logged in user, we need to wrap it in a `<section>`that renders conditionally when the `isLoggedIn` reference in the `useUserSession` is true.
```
```suggestion
Since it should only be shown to a logged in user, we need to wrap it in a `<section>` that renders conditionally when the `isLoggedIn` reference in the `useUserSession` is true.
```

gewenyu99 reviewed (approved):
Hey @evelinabe I just checked it out, amazing work.

We're down to just a few last comments, these comments are very small. The tutorial is already looking great.

I'm confident this will be the best tutorial in this repo when it's released.

I'll approve and mark accepted in case we don't get to it before Hacktoberfest ends, but please take a look at my remaining comments.

Thanks for you contribution to make Appwrite better for Nuxt developers!

gewenyu99 commented:
Hey,

Due to time constraints, I'm going to mark this PR `hacktoberfest-accepted` for now so you get DO's Hacktoberfest rewards. We'll continue to work with you on this issue for review and merge.

When it is merged, we'll contact you for Appwrite-specific Hacktoberfest swag.

Thanks for helping us improve Appwrite!

evelinabe commented:
> Hey @evelinabe I just checked it out, amazing work.
> 
> We're down to just a few last comments, these comments are very small. The tutorial is already looking great.
> 
> I'm confident this will be the best tutorial in this repo when it's released.
> 
> I'll approve and mark accepted in case we don't get to it before Hacktoberfest ends, but please take a look at my remaining comments.
> 
> Thanks for you contribution to make Appwrite better for Nuxt developers!

Hi @gewenyu99 !

Hope you've had a great time since Hacktober.
Thank you so much for your kind words, it means a lot!
I've finally fixed the requested changes for you.

gewenyu99 commented:
Hey there! There were a lot of big PRs during this Hacktoberfest, and we wanted to give everyone ample time to collaborate with our engineering team. If you were able to merge your PRs during October, amazing. If it’s still not merged, don’t worry about it either. Either way, we’ve got your Hacktoberfest swag minted and ready to ship. 

Please comment with your Discord username here so we can contact you about your shipping information to deliver your Hacktoberfest swag.
